src/main/java/htsjdk/variant/variantcontext/JEXLMap.java

```diff
@@ -11,152 +12,188 @@
             jexl.put(exp, value == null ? false : value);
-        } catch (Exception e) {
+        } catch (final JexlException e) {
+            // TODO: is this decision the desired result?
             // if exception happens because variable is undefined (i.e. field in expression is not present), evaluate to FALSE
+            jexl.put(exp,false);
+            // todo for a todo (must resolve both in PR review rounds): should I remove the todo and code below?
             // todo - might be safer if we explicitly checked for an exception type, but Apache's API doesn't seem to have that ability
-            if (e.getMessage() != null && e.getMessage().contains("undefined variable"))
-                jexl.put(exp,false);
-            else
-                throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s", exp.name, (e.getMessage() == null ? "no message" : e.getMessage())));
+            // if exception happens because variable is undefined (i.e. field in expression is not present), evaluate to FALSE
+            throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s",
+                    exp.name,
+                        /*(e.getMessage() == null ? */"no message" /*: e.getMessage())*/));
+        }
+    }
+
+    /**
+     * Create the internal JexlContext, only when required.
+     * This code is where new JEXL context variables should get added.
+     */
+    private void createContext() {
+        if ( vc == null ) {
+            jContext = new MapContext(Collections.emptyMap());
+        } else {
+            jContext = (g==null) ? new VariantJEXLContext(vc) : new GenotypeJEXLContext(vc, g);
         }
     }
 
     /**
-     * helper function: adds the list of attributes to the information map we're building
+     * TODO: function is not used, remove or not? resolve in PR review rounds.
+     * adds the list of attributes to the information map we're building
      * @param infoMap the map
      * @param attributes the attributes
      */
-    private static void addAttributesToMap(Map<String, Object> infoMap, Map<String, ?> attributes ) {
+    private static void addAttributesToMap(final Map<String, Object> infoMap, final Map<String, ?> attributes ) {
         for (Entry<String, ?> e : attributes.entrySet()) {
             infoMap.put(e.getKey(), String.valueOf(e.getValue()));
         }
     }
 
-    public Boolean put(JexlVCMatchExp jexlVCMatchExp, Boolean aBoolean) {
-        return jexl.put(jexlVCMatchExp,aBoolean);
-    }
-
-    public void putAll(Map<? extends JexlVCMatchExp, ? extends Boolean> map) {
-        jexl.putAll(map);
-    }
-
     // //////////////////////////////////////////////////////////////////////////////////////
+    // TODO: is the following comment still true? must resolve in PR review rounds
+    // TODO: is not supported, any other way to resolve at compile time rather than blow up in your face?
```

Comment:
These are fine I think.  There isn't a good way to convert these to a compile errors since java doesn't have separate interfaces for immutable collections.  It's pretty standard to implement some collection functions to just throw `UnsupportedOperationException`